### PR TITLE
Avoid fetching all stats metadata when there are no entities on the energy panel

### LINKED ISSUE #15589: Energy will request metadata for all entities when there are no matching entities
### Checklist

- [X] I have updated to the latest available Home Assistant version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.

### Describe the issue you are experiencing

Energy fires off a call with an empty array when there are no matching entities 

### Describe the behavior you expected

It should not fire off the call at all

### Steps to reproduce the issue

1.
2.
3.
...

Open energy panel with no matching ids

### What version of Home Assistant Core has the issue?

Dev

### What was the last working version of Home Assistant Core?

_No response_

### In which browser are you experiencing the issue with?

_No response_

### Which operating system are you using to run this browser?

_No response_

### State of relevant entities

_No response_

### Problem-relevant frontend configuration

_No response_

### Javascript errors shown in your browser console/inspector

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by bdraco**
## Proposed change

Avoid fetching all stats metadata when there are no entities. This is effectively the same issue as #15428 for metadata with less impact because it a smaller dataset being fetched.

I adjusted the code to fetch all the data at once since it is not dependent.

fixes #15589

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/data/energy.ts [hunk 1]
```diff
@@ -341,9 +339,8 @@ const getEnergyData = async (
   end?: Date,
   compare?: boolean
 ): Promise<EnergyData> => {
-  const [configEntries, entityRegistryEntries, info] = await Promise.all([
+  const [configEntries, info] = await Promise.all([
     getConfigEntries(hass, { domain: "co2signal" }),
-    subscribeOne(hass.connection, subscribeEntityRegistry),
     getEnergyInfo(hass),
   ]);
```
Comment (bdraco) [inline src/data/energy.ts:346]: These could probably be memorized so when you switch dates its not fetched again but I didn't want to make this any bigger.

File: src/data/energy.ts [hunk 2]
```diff
@@ -352,15 +349,15 @@ const getEnergyData = async (
     : undefined;
 
   let co2SignalEntity: string | undefined;
-
   if (co2SignalConfigEntry) {
-    for (const entry of entityRegistryEntries) {
-      if (entry.config_entry_id !== co2SignalConfigEntry.entry_id) {
+    for (const entityId of Object.keys(hass.entities)) {
```
Comment (bramkragten) [inline src/data/energy.ts:353]:
```suggestion
    for (const entityReg of Object.values(hass.entities)) {
```

File: src/data/energy.ts [hunk 3]
```diff
@@ -352,15 +349,15 @@ const getEnergyData = async (
...
+      const entity = hass.entities[entityId];
```
Comment (bramkragten) [inline src/data/energy.ts:354]:
```suggestion
    for (const entity of Object.Values(hass.entities)) {
```

File: src/data/energy.ts [hunk 4]
```diff
@@ -352,15 +349,15 @@ const getEnergyData = async (
...
+      if (entity.platform !== "co2signal") {
         continue;
       }
 
       // The integration offers 2 entities. We want the % one.
-      const co2State = hass.states[entry.entity_id];
+      const co2State = hass.states[entityId];
```
Comment (bramkragten) [inline src/data/energy.ts:360]:
```suggestion
      const co2State = hass.states[entity.entity_id];
```

File: src/data/energy.ts [hunk 5]
```diff
@@ -405,34 +402,35 @@ const getEnergyData = async (
...
+        startMinHour,
+        end,
+        energyStatIds,
+        period,
+        energyUnits,
+        ["sum"]
+      )
+    : Promise.resolve({});
+  const _waterStats = waterStatIds.length
+    ? await fetchStatistics(
```
Comment (bramkragten) [inline src/data/energy.ts:417]:
```suggestion
    ? fetchStatistics(
```

File: src/data/energy.ts [hunk 6]
```diff
@@ -405,34 +402,35 @@ const getEnergyData = async (
...
+    ? fetchStatistics(
+        hass!,
+        startMinHour,
+        end,
+        energyStatIds,
+        period,
+        energyUnits,
+        ["sum"]
+      )
+    : Promise.resolve({});
```
Comment (bramkragten) [inline src/data/energy.ts:415]:
```suggestion
    : {};
```

File: src/data/energy.ts [hunk 7]
```diff
@@ -405,34 +402,35 @@ const getEnergyData = async (
...
+    ? await fetchStatistics(
+        hass!,
+        startMinHour,
+        end,
+        waterStatIds,
+        period,
+        waterUnits,
+        ["sum"]
+      )
+    : Promise.resolve({});
```
Comment (bramkragten) [inline src/data/energy.ts:426]:
```suggestion
    : {};
```

File: src/data/energy.ts [hunk 8]
```diff
@@ -494,6 +492,37 @@ const getEnergyData = async (
     }
   }
 
+  const statsMetadata: Record<string, StatisticsMetaData> = {};
+  const _getStatisticMetadata: Promise<StatisticsMetaData[]> = allStatIDs.length
+    ? getStatisticMetadata(hass, allStatIDs)
+    : Promise.resolve([]);
```
Comment (bramkragten) [inline src/data/energy.ts:498]:
```suggestion
    : [];
```

File: src/data/energy.ts [hunk 9]
```diff
@@ -494,6 +492,37 @@ const getEnergyData = async (
     }
   }
 
+  const statsMetadata: Record<string, StatisticsMetaData> = {};
+  const _getStatisticMetadata: Promise<StatisticsMetaData[]> = allStatIDs.length
```
Comment (bramkragten) [inline src/data/energy.ts:496]:
```suggestion
  const _getStatisticMetadata: Promise<StatisticsMetaData[]> | StatisticsMetaData[] = allStatIDs.length
```

File: src/data/energy.ts [hunk 10]
```diff
@@ -443,38 +441,38 @@ const getEnergyData = async (
...
   }
 
-  let fossilEnergyConsumption: FossilEnergyConsumption | undefined;
-  let fossilEnergyConsumptionCompare: FossilEnergyConsumption | undefined;
-
+  let _fossilEnergyConsumption:
+    | Promise<undefined>
+    | Promise<FossilEnergyConsumption> = Promise.resolve(undefined);
+  let _fossilEnergyConsumptionCompare:
+    | Promise<undefined>
```
Comment (bramkragten) [inline src/data/energy.ts:472]:
```suggestion
    | undefined
```

File: src/data/energy.ts [hunk 11]
```diff
@@ -443,38 +441,38 @@ const getEnergyData = async (
...
+        ["sum"]
+      );
+    }
   }
 
-  let fossilEnergyConsumption: FossilEnergyConsumption | undefined;
-  let fossilEnergyConsumptionCompare: FossilEnergyConsumption | undefined;
-
+  let _fossilEnergyConsumption:
+    | Promise<undefined>
```
Comment (bramkragten) [inline src/data/energy.ts:469]:
```suggestion
    | undefined
```

File: src/data/energy.ts [hunk 12]
```diff
@@ -443,38 +441,38 @@ const getEnergyData = async (
...
+    | Promise<FossilEnergyConsumption> = Promise.resolve(undefined);
```
Comment (bramkragten) [inline src/data/energy.ts:470]:
```suggestion
    | Promise<FossilEnergyConsumption>;
```

File: src/data/energy.ts [hunk 13]
```diff
@@ -443,38 +441,38 @@ const getEnergyData = async (
...
+  let _fossilEnergyConsumptionCompare:
+    | Promise<undefined>
+    | Promise<FossilEnergyConsumption> = Promise.resolve(undefined);
```
Comment (bramkragten) [inline src/data/energy.ts:473]:
```suggestion
    | Promise<FossilEnergyConsumption>;
```

**Comment by bdraco** [comment]
Will rework after breakfast. Thanks

**Comment by bdraco** [comment]
Retested, should be good to go now.

**Review by bramkragten — APPROVED** [review]

**Pull request merged**